ray_on_golem/server/cluster/nodes.py

```diff
@@ -0,0 +1,381 @@
+
+        self._cluster = cluster
+        self._golem_service = golem_service
+        self._manager_stack = manager_stack
+
+        self._sidecars = self._prepare_sidecars()
+
+        with self.ssh_public_key_path.open() as f:
+            self._ssh_public_key_data = f.readline().strip()
+
+    def __str__(self) -> str:
+        return self.node_id
+
+    def get_data(self) -> NodeData:
+        return NodeData.parse_obj(self)
+
+    @property
+    def activity(self) -> Optional[Activity]:
+        return self._activity
+
+    def get_warning_messages(self) -> List[str]:
+        warnings = super().get_warning_messages()
+
+        for sidecar in self._sidecars:
+            warnings.extend(sidecar.get_warning_messages())
+
+        return warnings
+
+    def schedule_start(self) -> None:
+        if self._start_task and not self._start_task.done():
+            logger.info(f"Not scheduling start `%s` node, as it's already scheduled", self)
+            return
+
+        self._start_task = create_task_with_logging(
+            self.start(),
+            trace_id=get_trace_id_name(self, "scheduled-start"),
+        )
+
+    async def start(self) -> None:
+        if self.state in (NodeState.running, NodeState.terminating):
+            logger.info(f"Not starting `%s` node, as it's already running or stopping", self)
+            return
+
+        logger.info("Starting `%s` node...", self)
+
+        self.state = NodeState.pending
+
+        try:
+            while True:
+                try:
+                    activity, ip, ssh_proxy_command = await self._create_activity()
+                except ManagerException:
+                    # ManagerException have non-recoverable severity
+                    raise
+                except (GolemException, RayOnGolemError) as e:
+                    msg = "Failed to create activity, retrying."
+                    error = f"{type(e).__module__}.{type(e).__name__}: {e}"
+                    self._add_state_log(f"{msg} {error=}")
+                    logger.warning(msg)
+                    logger.debug(msg, exc_info=True)
+                else:
+                    break
+        except Exception as e:
+            self.state = NodeState.terminated
+            self._add_state_log(
+                f"Failed to create activity: {type(e).__module__}.{type(e).__name__}: {e}"
+            )
+
+            logger.info("Starting `%s` node failed!", self)
+
+            return
+
+        self._print_ssh_command(ip, ssh_proxy_command)
+
+        self.state = NodeState.running
+        self.internal_ip = ip
+        self.ssh_proxy_command = ssh_proxy_command
+
+        self._activity = activity
+
+        await self._start_sidecars()
+
+        logger.info("Starting `%s` node done", self)
+
+    async def stop(self) -> None:
+        if self.state in (NodeState.terminating, NodeState.terminated):
+            logger.info(f"Not stopping `%s` node, as it's already stopped", self)
```

Comment:
same here, could be simplified to: 

```suggestion
            logger.info(f"Node `%s` already terminated.", self)
```